documentation/specs/proposed/BuildCheck.md

```diff
@@ -0,0 +1,405 @@
+
+# BuildCheck - Design Spec
+
+AKA MSBuild Analyzers
+
+The feature is meant to help customers to improve and understand quality of their MSBuild scripts via rules violations reporting. It will allow MSBuild to gradually roll out additional rules, as users will be capable to configure their opt-in and severity of reports – preventing unwanted build breakages. And to equip powerusers to roll out their own quality checks – whether for general community or internal enterprise usage.
+
+# Table of Contents
+
+- [Terminology](#terminology)
+- [North Star / Longer-term vision](#north-star-longer-term-vision)
+- [Scope of initial iteration](#scope-of-initial-iteration)
+- [User Experience](#user-experience)
+   * [Running / UX](#running-ux)
+      + [Inbox Analyzers](#inbox-analyzers)
+      + [Live Build](#live-build)
+      + [Binlog Replay mode](#binlog-replay-mode)
+   * [Configuration](#configuration)
+      + [User Configurable Options](#user-configurable-options)
+         - [Enablement](#enablement)
+         - [Severity](#severity)
+         - [Scope of Analysis](#scope-of-analysis)
+      + [Sample configuration](#sample-configuration)
+   * [Analyzers and Rules Identification](#analyzers-and-rules-identification)
+   * [Custom Analyzers Authoring](#custom-analyzers-authoring)
+      + [Implementation](#implementation)
+         - [Analyzer declaration](#analyzer-declaration)
+         - [Rules declaration](#rules-declaration)
+         - [Standardized configuration declaration](#standardized-configuration-declaration)
+         - [Custom configuration declaration](#custom-configuration-declaration)
+         - [Compatibility](#compatibility)
+      + [Testing and Debugging](#testing-and-debugging)
+      + [Packaging](#packaging)
+   * [Acquisition of custom analyzers](#acquisition-of-custom-analyzers)
+
+
+
+# Terminology
+
+* **Analyzer** – unit of execution (single implementing class), can host multiple rules. 
+* **Rule** – Single violation type, with single unique code (`“MSB123: Redefining built-in target”`). 
```

Comment:
Please don't use MSB prefix in examples, we want people to do something distinct (so we don't get blamed :-P).

Reply:
Good catch.

100%